### ISSUE TITLE
prototype locking service for write conflict prevention

### ISSUE THREAD

**Issue by larkin** [nasa/openmct#136]
It seems that there is a need for a locking service which prevents objects from being mutated or deleted while another action is locking them.

An initial version of the service would provide a framework for locking objects in the context of a single user application, preventing users from taking actions that would conflict with async operations they are currently executing.

A second version of the service would use the same interface but depend on a back-end service to manage locking, so that it will work in a multi-user environment.  This may provide us with a way to remove our dependence on elasticsearch for object versioning (but unfortunately, add a different dependency).
### `lockService.lock(objectInstance, reason, waitFor)`

Obtain a lock on `objectInstance` for `reason`.
- **objectInstace** a domainObject to lock.
- **reason** a string describing the purpose of the lock.  
- **waitFor** integer, if provided, will continuously try to obtain a lock on the object until `waitFor` seconds have passed.

Returns a promise. If a lock is obtained, the promise is resolved with an `unlock` function which should be called when the lock is no longer required.  If a lock cannot be obtained, the promise is rejected with a message indicating the reason why the lock cannot be obtained (aka the reason of the pre-existing lock).

Locking an object requires that none of the objects in it's contextual path are locked. (locking a parent prevents locks on children until parent is unlocked).

example usage:

``` javascript
lockService.lock(object, 'so I can do something with it')
    .then(storeUnlockCallback, informUserOfConflict)
    .then(doActionWithObject)
    .then(unlockObject);
```
## Implementation ideas/questions
- I don't think mutation should obtain it's own locks-- it would be hard to pass meaningful "reasons" for a lock if that was done.
- that said, perhaps we require mutation to be passed a lock object or something that ensures mutation isn't being called without a lock.  Or do we just make it very clear in the developer guide?
- is it necessary to have locks time out after a certain period of time?  Should this be configurable in the lock call or not?  If locks can time out, how to we ensure that long running async calls don't take actions after a lock as timed out?  (I think it's best to not have locks timeout).
- if you wait for a lock on an object, and the object is deleted by it's other lock, well, how do you resolve this?

@VWoeltjen I don't believe this is a must-implement concept, but just wanted to throw it out as it may pertain to other discussions.  It seems that a per-client lock service could be useful for dealing with async operations, and a lightweight server implementation may help when working with persistence providers that don't provide conflict resolution.

**Comment by VWoeltjen** [comment]
I think this is a worthwhile approach to explore, but I will say that locking as an approach was tried with nasa/mct and became a sinkhole for engineering effort. It never worked well enough to ensure data integrity, so it didn't solve what it was meant to solve, and it also introduced new problems (performance artifacts, what to do when a lock doesn't get released, and so on.) 

This is also a little apples-and-oranges, because in the desktop client locking was also a user-facing feature (you would lock an object to start editing it) so there were usability issues with how cumbersome that was.

I mention this because if/when we go down that road, we should (a) look back over information from JIRA/Confluence about the problems that were encountered, (b) be prepared to meet resistance from stakeholders. I do agree that as the complexity of the operations we make available and the size of the object sets that people are working with both continue to scale, we will need to a good general solution for preserving data integrity amidst concurrent tasks.

**Comment by larkin** [comment]
This is definitely separate from MCT style locking; the lockService should never be used to hold a lock while waiting for user input.  Otherwise it's going to be totally broken in the ways you described.

The purpose here is to start building a system which ensures data integrity without compromising interface design or user stories.  Two users editing the same object should see their changes merged safely-- if I change the name and you change the period of a sine wave, the end result should include both changes.  If two people renamed an object two different names and only one of those names persisted, would anyone really be confused?  No-- the name still would have changed, and whoever's name was overwritten would just think "someone else must have changed it after me," which is, truth be told, exactly what happened.

Bottom line: it should ***just work.***

It is possible to hide 100% of conflict resolution from the end user if we use a locking system in combination with a high-fidelity mutation tracking/application system.  I am fully behind not making this a user-interface concern, ever.  

Longer term, if we track all object mutations in a server-side journal, we have full, complete accountability and fine-grained point-in-time backup and restore capability, so if a conflict resolution goes wrong, we can replay it and make it go right.

**Comment by larkin** [comment]
closing as wontfix at this time.
